docker/Dockerfile.deb_debug

```diff
@@ -6,6 +6,7 @@ FROM debian:9-slim
 # Do not add more stuff to this list that isn't small or critically useful.
 # If you occasionally need something on the container do
 # sudo apt-get update && apt-get whichever
+# hadolint ignore=DL3005,DL3008
```

Comment:
didn't we make it a big point to run `apt-get upgrade` before, and now hadolint is saying that is bad? Should we `upgrade` or not?

Reply:
I do agree, hadolint / best practice suggest to pin everything.
If we go this direction, (and I think we should), it is going to be a very tedious job.

Reply:
I don't think we want to `apt-get upgrade` since that is trying to bring up to date all the previously installed bits. It makes sense to have the underlying container be rebuilt to get these changes (whether it's a security fix or not).

Doing the `apt-get update` gets an updated list of packages. If we are installing a package on top of the base image, it makes sense to do the _update_ to make sure we are getting the latest and any fixes that might need to be pulled in. Ex. There is a CVE for `curl` and we want the latest _curl_ it should come in with a  _update_ and _install_. This does have an issue if we want repeatability in an image (it's hard to do this as the underlying image may change). Ex. We use ubuntu and a specific curl version. We rebuild and may get a different ubuntu but the same curl.

So, I think we want to make sure to do the _update_ and the _install_, but not an _upgrade_. My votes has been to not pin things, because then you must make a Dockerfile change to pin to the next version before you can rebuild. Ex with _curl_ again. If there is a need to rebuild to get a fix, we would need to update the Dockerfile if it was pinned and rebuild. This precludes any real automation. If it's not pinned, we can simply run the Docker build again and get a new image that satisfies the CVE.

Reply:
I should add, that pinning can be helpful if we have a specific version that we need for compatibility. Something like a version of _kind_ or _Golang_.

Reply:
@ericvn I think I'd like to phrase the problem with the following questions:
_"What can we **observe** today on the software supply chain of an Istio release?
Is there any way for an Istio user, to know about the software ran in production?"_

To answer these today, users are running vulnerability scanners, and reporting what they're seeing, that's what happened with #14220, and from our perspective, we're a bit too "reactive".

What a more proactive situation would be, is the following:
Ideally, when a vulnerability is found in a software, reported and filed as a CVE, we could have a security linter that verify that scans our Dockerfiles and report what is  affected by a vulnerability just by version matching; realistically, we could even expect GitHub to do that job at some point  (they're already scanning for other dep files, and my guess is that they'll do it for Dockerfile too).
But for this to happen, we need to pin the version**s** (both the image version **AND** software versions), then we can continuously assess the security of the supply chain.

Another part of the problem is not only making sure we have the last update package (like your curl example), but also to actually make sure that the image (e.g. ubuntu, but we have other sources too), actually have built an updated curl for an hypothetically disclosed CVE. What if they have not, or not yet? Depending on a best effort is not enough (and sounds an awful lot like security by obscurity), and we might want to resort to build the new software in the image ourselves if the supply chain does not provide it. Again, this can only be done by knowing precisely what we are running.

Just for the record: my opinion at first was "let's just run an update + upgrade and we're good, running the latest things", but after the last few CVEs, I actually think that we're too reactive about security and it bites us every time we got a report (that is considered like public disclosure from day-1 and forces us to release in a hurry).

Reply:
@fpesce I agree with what you are saying, but I have also been under the gun in the past to make sure what's available is free of vulnerabilities. I think part of the issue is what's being provided as a service vs provided as a product need to be handled differently. It's easier to control what's deployed when you are the service then when you are a product and people can pull old and unsupported images, etc.

I haven't had much visibility to the Istio CVEs (and that's not necessarily a bad thing), so it's hard for me to make a call there. I defer to you and others that have had to do the reactionary items.

Reply:
pinning is the only legitimate approach for docker files.  Unfortunately, docker does not permit a range of releases (e.g. 1.17.1-1.19) in which case, a hard pin wouldn't be necessary.  Using latest results in all sorts of problems, most importantly, you don't know which version of the software has been built :)